src/astUtils/visitors.ts

```diff
@@ -1,14 +1,77 @@
+/* eslint-disable no-bitwise */
 import { CancellationToken } from 'vscode-languageserver';
-import { Statement, Body, AssignmentStatement, Block, ExpressionStatement, CommentStatement, ExitForStatement, ExitWhileStatement, FunctionStatement, IfStatement, IncrementStatement, PrintStatement, GotoStatement, LabelStatement, ReturnStatement, EndStatement, StopStatement, ForStatement, ForEachStatement, WhileStatement, DottedSetStatement, IndexedSetStatement, LibraryStatement, NamespaceStatement, ImportStatement } from '../parser/Statement';
-import { Expression } from '../parser/Expression';
-import { isExpression, isBlock, isIfStatement } from './reflection';
+import { Statement, Body, AssignmentStatement, Block, ExpressionStatement, CommentStatement, ExitForStatement, ExitWhileStatement, FunctionStatement, IfStatement, IncrementStatement, PrintStatement, GotoStatement, LabelStatement, ReturnStatement, EndStatement, StopStatement, ForStatement, ForEachStatement, WhileStatement, DottedSetStatement, IndexedSetStatement, LibraryStatement, NamespaceStatement, ImportStatement, ClassStatement, ClassMethodStatement, ClassFieldStatement } from '../parser/Statement';
+import { AALiteralExpression, ArrayLiteralExpression, BinaryExpression, CallExpression, CallfuncExpression, DottedGetExpression, EscapedCharCodeLiteralExpression, Expression, FunctionExpression, GroupingExpression, IndexedGetExpression, LiteralExpression, NamespacedVariableNameExpression, NewExpression, SourceLiteralExpression, TaggedTemplateStringExpression, TemplateStringExpression, TemplateStringQuasiExpression, UnaryExpression, VariableExpression, XmlAttributeGetExpression } from '../parser/Expression';
+import { isExpression, isStatement } from './reflection';
 
 
 /**
- * Create a filtered visitor for use with `editStatements`
+ * Walks the statements of a block and descendent sub-blocks, and allow replacing statements
  */
-export function createStatementVisitor(
+export function walkStatements(
+    statement: Statement,
+    visitor: (statement: Statement, parent: Statement) => Statement | void,
+    cancel?: CancellationToken
+): void {
+    statement.walk(visitor as any, {
+        walkMode: WalkMode.visitStatements,
+        cancel: cancel
+    });
+}
+
+
+export type WalkVisitor = <T = Statement | Expression>(stmtExpr: Statement | Expression, parent: Statement | Expression) => void | T;
+
+/**
+ * A helper function for Statement and Expression `walkAll` calls.
+ */
+export function walk<T>(keyParent: T, key: keyof T, visitor: WalkVisitor, options: WalkOptions, parent?: Expression | Statement) {
+    //stop processing if canceled
+    if (options.cancel?.isCancellationRequested) {
+        return;
+    }
+
+    //the object we're visiting
+    let element: any = keyParent[key];
+    if (!element) {
+        return;
+    }
+
+    //notify the visitor of this element
+    if (
+        (isExpression(element) && options.walkMode & WalkModeInternal.visitExpressions) ||
```

Comment:
@elsassph do you have any suggestions for a more efficient way to handle the walk/visit logic? Right now, we have to determine each element's type before we can know for certain that we should notify the visitor.

Reply:
Ouch.
This reminds me that `isExpresion` isn't reliable, now that you added `walk` functions to some statements.

Most efficient would be that all expressions and statements have a `walkMode` property, so you could just write:
```
if (element.walkMode & opens.walkMode) {...}
```

Reply:
Oh, I like that approach. And we can set that in the superclass constructor so it doesn't clutter the Statements and Expressions.